aya/src/programs/extension.rs

```diff
@@ -0,0 +1,186 @@
+use std::{
+    convert::TryFrom,
+    os::unix::prelude::{AsRawFd, RawFd},
+};
+use thiserror::Error;
+
+use object::Endianness;
+
+use crate::{
+    generated::{bpf_attach_type::BPF_CGROUP_INET_INGRESS, bpf_prog_type::BPF_PROG_TYPE_EXT},
+    obj::btf::BtfKind,
+    programs::{load_program, LinkRef, ProgramData, ProgramError},
+    sys::{self, bpf_link_create},
+    Btf,
+};
+
+use super::{FdLink, Program};
+
+/// The type returned when loading or attaching an [`Extension`] fails
+#[derive(Debug, Error)]
+pub enum ExtensionError {
+    #[error("target BPF program does not have BTF loaded to the kernel")]
+    NoBTF,
+}
+
+/// A program used to extend existing BPF programs
+///
+/// [`Extension`] programs can be loaded to replace a global
+/// function in a program that has already been loaded.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 5.9
+///
+/// # Examples
+///
+/// ```no_run
+/// use aya::{BpfLoader, programs::{Xdp, XdpFlags, Extension, ProgramFd}};
+/// use std::convert::TryInto;
+///
+/// let mut bpf = BpfLoader::new().load_file("app.o")?;
+/// let prog: &mut Xdp = bpf.program_mut("main").unwrap().try_into()?;
+/// prog.load()?;
+/// prog.attach("eth0", XdpFlags::default())?;
+///
+/// let prog_fd = prog.fd().unwrap();
+/// let ext: &mut Extension = bpf.program_mut("extension").unwrap().try_into()?;
+/// ext.load(prog_fd, "function_to_replace")?;
+/// ext.attach()?;
+/// Ok::<(), aya::BpfError>(())
+/// ```
+#[derive(Debug)]
+#[doc(alias = "BPF_PROG_TYPE_EXT")]
+pub struct Extension {
+    pub(crate) data: ProgramData,
+}
+
+impl Extension {
+    /// Loads the program inside the kernel.
+    ///
+    /// See also [`Program::load`](crate::programs::Program::load).
+    pub fn load<T: AsRawFd>(&mut self, program: T, func_name: &str) -> Result<(), ProgramError> {
```

Comment:
The docs need to explain what program and func_name do

Reply:
I would say something like:

```
Loads the program extension.

Prepares the code included in the extension to replace the code of the function `func_name` within the eBPF program represented by the `program` file descriptor. The extension code will be loaded but inactive until it's attached.

# Example 
...

```

Reply:
It should also explicitly say whether you can replace a top level program or not, eg:

```
#[xdp]
fn outer(ctx: XdpContext) -> XdpAction {
	inner(ctx)
}

fn inner(ctx: XdpContext) -> XdpAction {
	...
}
```

```
let outer = ...;
let ext = ...;

ext.load(outer, "outer")? // <- does this work?
```